pkg/operator/operator.go

```diff
@@ -12,140 +12,311 @@ distributed under the License is distributed on an "AS IS" BASIS,
-	waitCluster      sync.WaitGroup
-	factory          client.ConnectionFactory
-	useAllDevices    bool
+	Namespace   string
+	MasterHost  string
+	clientset   *kubernetes.Clientset
+	waitCluster sync.WaitGroup
+	factory     client.ConnectionFactory
+	stopChMap   map[string]chan struct{}
+	clusters    map[string]*Cluster
+	kubeHttpCli *http.Client
+	// Kubernetes resource version of the clusters
+	clusterRVs map[string]string
 }
 
-func New(namespace string, factory client.ConnectionFactory, clientset *kubernetes.Clientset, containerVersion string, useAllDevices bool) *Operator {
+func New(host, namespace string, factory client.ConnectionFactory, clientset *kubernetes.Clientset) *Operator {
 	return &Operator{
-		Namespace:        namespace,
-		factory:          factory,
-		clientset:        clientset,
-		containerVersion: containerVersion,
-		useAllDevices:    useAllDevices,
+		Namespace:  namespace,
+		MasterHost: host,
+		factory:    factory,
+		clientset:  clientset,
+		clusters:   make(map[string]*Cluster),
+		clusterRVs: make(map[string]string),
+		stopChMap:  map[string]chan struct{}{},
 	}
 }
 
 func (o *Operator) Run() error {
-
-	// Start the mon pods
-	m := mon.New(o.Namespace, o.factory, o.containerVersion)
-	cluster, err := m.Start(o.clientset)
-	if err != nil {
-		return fmt.Errorf("failed to start the mons. %+v", err)
+	var watchVersion string
+	var err error
+	for {
+		watchVersion, err = o.initResources()
+		if err == nil {
+			break
+		}
+		logger.Errorf("failed to create tpr. %+v. retrying...", err)
+		<-time.After(initRetryDelay)
 	}
 
-	a := api.New(o.Namespace, o.containerVersion)
-	err = a.Start(o.clientset, cluster)
+	// watch for changes to the rook clusters
+	return o.watchTPR(watchVersion)
+}
+
+func (o *Operator) initResources() (string, error) {
+	httpCli, err := newHttpClient()
 	if err != nil {
-		return fmt.Errorf("failed to start the REST api. %+v", err)
+		return "", fmt.Errorf("failed to get tpr client. %+v", err)
 	}
+	o.kubeHttpCli = httpCli.Client
 
-	// Start the OSDs
-	osds := osd.New(o.Namespace, o.containerVersion, o.useAllDevices)
-	err = osds.Start(o.clientset, cluster)
+	err = o.createTPR()
 	if err != nil {
-		return fmt.Errorf("failed to start the osds. %+v", err)
+		if k8sutil.IsKubernetesResourceAlreadyExistError(err) {
+			// TPR has been initialized before. We need to recover existing cluster.
+			watchVersion, err := o.findAllClusters()
+			if err != nil {
+				return "", fmt.Errorf("failed to find clusters. %+v", err)
+			}
+			return watchVersion, nil
+		}
 	}
 
-	// Start the object store
-	r := rgw.New(o.Namespace, o.containerVersion, o.factory)
-	err = r.Start(o.clientset, cluster)
+	return "0", nil
+}
+
+func (o *Operator) watchTPR(watchVersion string) error {
+	logger.Infof("start watching rook tpr: %s", watchVersion)
+	defer func() {
+		for _, stop := range o.stopChMap {
+			close(stop)
+		}
+		o.waitCluster.Wait()
+	}()
+
+	eventCh, errCh := o.watch(watchVersion)
+
+	go func() {
```

Comment:
I probably dont understand much about how we are leveraging TPR but is there a need to watch it? Doesnt kubernetes handle that for us? If i create a Kind: RookCluster, the TPR will handle that. And if we need to configure it or updated it, we could leverage kubectl.